### ISSUE TITLE
Example DIALS input files to add CrystFEL's stream format to stills2mtz

### ISSUE THREAD

**Issue by marinegor** [Hekstra-Lab/careless#5]
Hi everyone, thanks for your awesome preprint & software!

Could you please share examples of input files for stills2mtz? As I understand, these should be:
    - DIALS experiment file
    - DIALS reflection table
I'd really appreciate human-readable format, i.e. .json and .refl, respectively. If you have files `thermolysin_xfel` (from `careless-examples`) in hands, they'd be the best option.

My idea is to try to implement CrystFEL's stream format conversion into .json and .refl, respecitvely, so that `careless` could run on non-DIALS still datasets, too. I hope to improve my (yet unpublished) data using `careless`, and really hope that a connection with CrystFEL's ecosystem would help `careless` to become more widely used in the SSX/SFX community!

**Comment by kmdalton** [comment]
Hi Egor,

Happy to see the interest in our code. A few thoughts:
 - I've no problem with uploading the json and reflection files I used in the manuscript to the `careless-examples` repo. I can do that this afternoon. For reference, the files are available from the [CXIDB](https://cxidb.org/id-81.html). However, you have to download the full archive which is ~20 GB. It makes sense to upload the individual run I used.
 - The only input format that `careless` currently supports is `mtz`. So you might be better off going directly from stream to `mtz`. 
   - You'll have to figure out how to annotate your `mtz` with the desired metadata. 
   - We've been using [reciprocalspaceship](https://github.com/Hekstra-Lab/reciprocalspaceship) for these sorts of tasks internally. In brief, it lets you interact with `mtz` files as though they were `pandas DataFrames`. 
 - Long term there are a few things you should be aware of regarding SFX
   - For XFEL data, you should always use the `--image-scale-key=` command line argument in order to fit per image scales. I think this is covered in the XFEL [example](https://github.com/Hekstra-Lab/careless-examples/blob/main/XFEL.md). Anecdotally, I found this helps even for single crystal XFEL data. 
   - The current implementation is limited to a single gpu which means you start to run out of memory after about 10 million reflection observations. The exact limit depends on how many metadata columns you include. Depending on how big your data are, `careless` might not be an appropriate tool for your application yet. I'm working on two strategies to alleviate this limitation
     - Variational models that can train *stochastically* by looking at one image at a time. I have a proof of concept for this, but it is still several months from being a technology. 
     - A distributed version of `careless` that uses multiple gpus. My initial read of the [tf.distribute.Strategy](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy?version=nightly) documentation suggests that this would be possible with some limited modifications to the model. I probably won't have the bandwidth to test this out until March/April.

**Comment by JBGreisman** [comment]
As Kevin mentioned, the internals of `careless` are built around being able to construct a `reciprocalspaceship.DataSet` object containing all the per-reflection data/metadata (this is a crystallographically-aware `pandas DataFrame`). 

If there are additional useful file formats that you think should be supported, I think that will be best handled in [`reciprocalspaceship`](https://github.com/Hekstra-Lab/reciprocalspaceship). Currently, `rs` directly supports MTZ files, Precognition files (Laue processing software), and CSV files (and any other simple "separator delimited" files). However, it is also possible to turn an ordinary `pandas.DataFrame` into a `DataSet` object, so that may make it possible to support other inputs using `pandas` I/O parsers (such as `pandas.read_json()`).

**Comment by kmdalton** [comment]
Hi @marinegor, I spun up a little [repo](https://github.com/Hekstra-Lab/stills2mtz) to demonstrate `stills2mtz`. Unfortunately, the DIALS reflection file format is binary and complicated, and I cannot really offer you a plain text version at the moment.

**Comment by marinegor** [comment]
Ok, that's super detailed, thanks everyone for the feedback (especially @kmdalton)

Point-by-point:
>However, you have to download the full archive which is ~20 GB
The full 20 Gb archive is not a problem anyway -- it takes ~ 1h to download it.
The full dataset, however, is a bit larger problem, but I'll probably download it anyway to use CrystFEL on it.

>So you might be better off going directly from stream to `mtz`
Indeed, thanks. As I can see from your repo stills2mtz, it's probably much simplier than I thought initially.

I'll get back to the thread soon, thanks a lot for the prompt reply!

**Comment by marinegor** [comment]
Hi everyone, the download is going well, I hope to finish all ~6 Tb of raw data from cxidb_id81 within 2 weeks.

Meanwhile, couple of questions regarding `careless` and `stills2mtz`:

1. Instead of modifying `stills2mtz` for CrystFEL, I decided to implement `reciprocalspaceship/io/crystfel.py` function to read CrystFEL's stream into a `rs.DataSet()` format, which seems in line with the overall library format. Reading labels `H, K, L, I, sigmaI` would be very straightforward, as well as supplying them with `ewald_offset` per reflection.
2. I suppose there is also a need for `BATCH` column, indicating the crystal's ID? You didn't mention it in the stills2mtz repo, but it seems right.
3. I do not fully understand, what are `xobs, yobs` labels in the `mtz` file, and how `caresless` uses them. From the `careless-examples/thermolysin_xfel/unmerged.mtz` I can see that `(xobs, yobs)` are roughly 194x185, which corresponds to a single CSPAD panel and thus are relative. But what are they for, and how are they useful during scaling?

**Comment by kmdalton** [comment]
1. This is great! Look forward to a pull request.
2. Yes, a `BATCH` column will be necessary for good scaling. 
3. I will look into this further. It is fully possible that I goofed and used the wrong column for `x/yobs`. My intention was to get the position of the reflections in real space not the position on the individual panel. For historical context, I adapted `stills2mtz` from a script that was used to process data from a monolithic detector. For CSPAD data, I probably should have used `xyzobs.mm.value` rather than `xyzobs.px.value`. I'll get back with an update after I confirm this. 

At any rate, the idea is that you want to give the model the opportunity to account for the location of the Bragg peak on the detector so that it can learn to correct for polarization or detector artifacts. I think there are a lot of roughly equivalent ways to accomplish this. One obvious way would be to supply the `S1` wavevector for each reflection. That sort of parameterization has have the advantage of being portable across detectors. Maybe consider outputting `S1_{X,Y,Z}` columns. Alternatively, you could use the position of the reflections within the whole detector system rather than just the panel. Does that make sense?

**Comment by marinegor** [comment]
> For CSPAD data, I probably should have used xyzobs.mm.value rather than xyzobs.px.value. I'll get back with an update after I confirm this.

Ok, I see. I understand that CSPAD's geometry might be a huge pain -- If there are alternatives, I'd strongly prefer not to dive there :)

> Maybe consider outputting S1_{X,Y,Z} columns.

That's simple enough. If you can modify `careless` to work with it, that should be a very easy thing to add.

> Alternatively, you could use the position of the reflections within the whole detector system rather than just the panel. Does that make sense?

Yes, it does. If you for some reason still prefer to work with the real space coordinates, there is a [cfelpyutils](https://github.com/lucagelisio/cfelpyutils) library that can help you with.

**Comment by kmdalton** [comment]
In principle it shouldn't matter whether you provide real space coords or S1. However, there are still a lot of open questions about what strategies are optimal for different sorts of data. 

As far as modifying `careless`, the only changes that should need to be made are at the level of the command line interface. The code can already learn nonlinear mappings between inputs and scales. So, the only thing that needs to change is the metadata key supplied by the user. 

I can confirm the pixel positions were per panel. This was very much not my intention. So, I'll redo the analysis with the S1 vectors and update the pre-print in the next release. Thanks for looking into this!

**Comment by marinegor** [comment]
Hi @kmdalton,

I suddenly realized that I can not work with the downloaded xtc files from cxidb_id81 -- CrystFEL works with either hdf or cbf files, and also requires CrystFEL's geometry format of the CSPAD detector. Do you, by any chance, have either conversion tools or already converted runs that you used in your preprint?

It would be nice to compare all merging appoaches (CrystFEL's `partialator`, `cxi.merge`, etc) with `careless` on the same data, rather than on different datasets. Alternatively, I could use any other public data from cxidb, but it's only the backup plan.

**Comment by kmdalton** [comment]
Hey @marinegor,

`cxidb 81` also includes the `DIALS` integration output in the form of
 - experimental geometry files (`json`) and
 - reflection tables (`pickle`)
 
That's what I have been working with. I don't really have the tooling to process the raw `xtc` files without considerable effort. I think it's also still quite an undertaking to do a good job integrating CSPAD data, and I would probably defer to collaborators for help there. I went with `cxidb 81` for the `careless` pre-print because I trust that the `DIALS` processing for that particular experiment is high quality. 

In the pre-print I just processed a single run (~3,000 images). I converted a single pair of `json` and `pickle` files into an `mtz` using that `stills2mtz` code. The resulting `mtz` is in the `careless-examples` folder. Can you convert that `mtz` into a format `partialator` understands? I have never run `CrystFEL`, can you give me an idea of what the most useful inputs would be? 

If you already have data which you've processed in `CrystFEL`, it would probably be pretty straightforward to get it into `careless`. All you'd need is to make an `mtz` file with the `S1` wavevector, `ewald offsets`, `I` and `SigI`. If you know how to get these data out of `CrystFEL` it'd be really easy to dump them to a compatible `mtz` with `reciprocalspaceship`. 

In short, I can make unmerged `mtz` files from DIALS output, but I cannot really help when it comes to getting the raw data into `CrystFEL`. I hope that makes sense. Let me know what I can do to help.

**Comment by marinegor** [comment]
Ok, I see.

Unfortunately, turning back from `json/pickle` formats to CrystFEL is too complicated, so I'd better use another dataset here.
I have some S-SAD data in hands, as well as a normal test lysozyme data -- I'll try both and see which one works.

**Comment by kmdalton** [comment]
Sounds good. Let me know if you hit any snags.

**Comment by kmdalton** [comment]
I am closing this issue, because I think we solved it over in `reciprocalspaceship`.
